pkg/collector/corechecks/system/winkmem/winkmem.go

```diff
@@ -0,0 +1,239 @@
+		TopPagedAllocsOutstanding:    KMemDefaultTopNum,
+		TopNonPagedAllocsOutstanding: KMemDefaultTopNum,
+	}
+	err := yaml.Unmarshal(initConfig, &cf)
+	if err != nil {
+		return err
+	}
+
+	w.config = cf
+	log.Infof("Winkmem config %v", w.config)
+	return nil
+}
+
+// Run executes the check
+func (w *KMemCheck) Run() error {
+	sender, err := aggregator.GetSender(w.ID())
+	if err != nil {
+		return err
+	}
+	spi, err := getpoolinfo()
+	if err != nil {
+		return err
+	}
+	// make a map to keep track of the indexes we actually want
+	// to send up (so we don't double count them)
+	tagmap := make(map[int]bool)
+	for i := 0; i < w.config.TopPagedBytes; i++ {
+		tagmap[spi.pagedPoolBytes[i].Value] = true
+	}
+	for i := 0; i < w.config.TopNonPagedBytes; i++ {
+		tagmap[spi.nonPagedPoolBytes[i].Value] = true
+	}
+	for i := 0; i < w.config.TopPagedAllocsOutstanding; i++ {
+		tagmap[spi.pagedPoolAllocsOutstanding[i].Value] = true
+	}
+	for i := 0; i < w.config.TopNonPagedAllocsOutstanding; i++ {
+		tagmap[spi.nonPagedPoolAllocsOutstanding[i].Value] = true
+	}
+	if len(w.config.SpecificTags) > 0 {
+		for _, ktag := range w.config.SpecificTags {
+			tagmap[spi.byKey[ktag]] = true
+		}
+	}
+
+	// now we have a list of indexes we actually want to store.  Walk them,
+	// and do it
+	for k, v := range tagmap {
+		// double sanity check, but should always be true
+		if v {
+			var tags = []string{}
+			pti := spi.spti.poolTags[k]
+
+			tags = append(tags, "kmemtag:"+string(pti.tag[:]))
+			log.Debugf("Logging mem tag %v", tags)
+
+			sender.Gauge("winkmem.paged_pool_bytes", float64(pti.pagedUsed), "", tags)
+			sender.Gauge("winkmem.nonpaged_pool_bytes", float64(pti.nonPagedUsed), "", tags)
+			sender.Gauge("winkmem.paged_allocs_outstanding", float64(pti.pagedAllocs-pti.pagedFrees), "", tags)
+			sender.Gauge("winkmem.nonpaged_allocs_outstanding", float64(pti.nonPagedAllocs-pti.nonPagedFrees), "", tags)
+		}
+	}
+	sender.Commit()
+	log.Debugf("Logged %v entries", len(tagmap))
+	return nil
+
+}
+
+type systemPooltag struct {
+	tag            [4]uint8
+	pagedAllocs    uint32
+	pagedFrees     uint32
+	pagedUsed      uint64
+	nonPagedAllocs uint32
+	nonPagedFrees  uint32
+	nonPagedUsed   uint64
+}
+type systemPooltagInformation struct {
+	count    uint32
+	padding  uint32
+	poolTags []systemPooltag
+}
+
+type systemPoolInformation struct {
+	spti                          systemPooltagInformation
+	byKey                         map[string]int
+	pagedPoolBytes                sortKeyList
+	nonPagedPoolBytes             sortKeyList
+	pagedPoolAllocsOutstanding    sortKeyList
+	nonPagedPoolAllocsOutstanding sortKeyList
+}
+type sortKey struct {
+	Key   uint64 // numerical value (i.e. allocs, frees, etc)
+	Value int    // index into pooltags array
+}
+type sortKeyList []sortKey
+
+func (p sortKeyList) Len() int { return len(p) }
+func (p sortKeyList) Swap(i, j int) {
+	l1, l2 := p[j], p[i]
+	p[j], p[i] = p[i], p[j]
+	p[i], p[j] = l1, l2
```

Comment:
This line is redundant. `p[i]` is already assigned value of `p[j]` on line 193 (same for `p[j]`).